`OneHotEncoder` causing shape misalignment in `CopulaGAN`, `CTGAN`, and `TVAE`

### Environment Details

Please indicate the following details about the environment in which you found the bug:

* RDT version: 0.5.2.dev0
* Python version: 3.8
* Operating System: macOS 11.5.1

### Error Description

After optimizing the OneHotEncoder Transformer, `CopulaGAN`, `CTGAN`, and `TVAE` start failing in SDV. The primary reason for this failure is caused by the line:

https://github.com/sdv-dev/CTGAN/blob/2dbbc18770d0dc17e7c5e83da6980ace424e9326/ctgan/data_transformer.py#L57

where DataTransformer access `ohe.dummies`. In PR #186, we altered the contents of `self.dummies` to not include NaN values (because `pd.Categorical` doesn't accept it as a category), which inconsequence makes us initialize weights of the the wrong dimensions in consequent parts of the code because we are missing the NaN value in `ohe.dummies`. 

### Steps to reproduce

Initialize any dataframe with missing values and it will crash

```python3
import pandas as pd
from sdv.tabular import CTGAN

data = pd.DataFrame({"category": ["1", "2", "3", None] * 4})
ctgan = CTGAN()
ctgan.fit(data)
```

the code crashes with the following message
```
AssertionError                            Traceback (most recent call last)
<ipython-input-8-8847fc3cae96> in <module>
----> 1 ctgan.fit(data)

~/Downloads/repos/SDV/sdv/tabular/base.py in fit(self, data)
    140             LOGGER.debug(
    141                 'Fitting %s model to table %s', self.__class__.__name__, self._metadata.name)
--> 142             self._fit(transformed)
    143 
    144     def get_metadata(self):

~/Downloads/repos/SDV/sdv/tabular/ctgan.py in _fit(self, table_data)
     55                     categoricals.append(field)
     56 
---> 57         self._model.fit(
     58             table_data,
     59             discrete_columns=categoricals

~/opt/anaconda3/envs/rdt/lib/python3.8/site-packages/ctgan/synthesizers/ctgan.py in fit(self, train_data, discrete_columns, epochs)
    296         train_data = self._transformer.transform(train_data)
    297 
--> 298         self._data_sampler = DataSampler(
    299             train_data,
    300             self._transformer.output_info_list,

~/opt/anaconda3/envs/rdt/lib/python3.8/site-packages/ctgan/data_sampler.py in __init__(self, data, output_info, log_frequency)
     37             else:
     38                 st += sum([span_info.dim for span_info in column_info])
---> 39         assert st == data.shape[1]
     40 
     41         # Prepare an interval matrix for efficiently sample conditional vector

AssertionError: 
```

### Steps to fix
1. revisit the variables in `OneHotEncoder` and let `dummies` contain `np.nan` when that's the case.
2. add a unit test to confirm `np.nan` is part of `dummies` when `dummy_na` is set to True.

--- a/rdt/transformers/categorical.py
+++ b/rdt/transformers/categorical.py
@@ -260,11 +260,11 @@ class OneHotEncodingTransformer(BaseTransformer):
     """
 
     dummies = None
-    dummy_na = None
-    num_dummies = None
-    dummy_encoded = False
-    indexer = None
-    decoder = None
+    _dummy_na = None
+    _num_dummies = None
+    _dummy_encoded = False
+    _indexer = None
+    _uniques = None
 
     def __init__(self, error_on_unknown=True):
         self.error_on_unknown = error_on_unknown
@@ -297,19 +297,19 @@ class OneHotEncodingTransformer(BaseTransformer):
         return data
 
     def _transform(self, data):
-        if self.dummy_encoded:
-            coder = self.indexer
-            codes = pd.Categorical(data, categories=self.dummies).codes
+        if self._dummy_encoded:
+            coder = self._indexer
+            codes = pd.Categorical(data, categories=self._uniques).codes
         else:
-            coder = self.dummies
+            coder = self._uniques
             codes = data
 
         rows = len(data)
-        dummies = np.broadcast_to(coder, (rows, self.num_dummies))
-        coded = np.broadcast_to(codes, (self.num_dummies, rows)).T
+        dummies = np.broadcast_to(coder, (rows, self._num_dummies))
+        coded = np.broadcast_to(codes, (self._num_dummies, rows)).T
         array = (coded == dummies).astype(int)
 
-        if self.dummy_na:
+        if self._dummy_na:
             null = np.zeros((rows, 1), dtype=int)
             null[pd.isnull(data)] = 1
             array = np.append(array, null, axis=1)
@@ -328,17 +328,17 @@ class OneHotEncodingTransformer(BaseTransformer):
         data = self._prepare_data(data)
 
         null = pd.isnull(data)
-        self.dummy_na = null.any()
-        self.dummies = list(pd.unique(data[~null]))
-        self.num_dummies = len(self.dummies)
-        self.indexer = list(range(self.num_dummies))
-        self.decoder = self.dummies.copy()
+        self._uniques = list(pd.unique(data[~null]))
+        self._dummy_na = null.any()
+        self._num_dummies = len(self._uniques)
+        self._indexer = list(range(self._num_dummies))
+        self.dummies = self._uniques.copy()
 
         if not np.issubdtype(data.dtype, np.number):
-            self.dummy_encoded = True
+            self._dummy_encoded = True
 
-        if self.dummy_na:
-            self.decoder.append(np.nan)
+        if self._dummy_na:
+            self.dummies.append(np.nan)
 
     def transform(self, data):
         """Replace each category with the OneHot vectors.
@@ -375,7 +375,7 @@ class OneHotEncodingTransformer(BaseTransformer):
             data = data.reshape(-1, 1)
 
         indices = np.argmax(data, axis=1)
-        return pd.Series(indices).map(self.decoder.__getitem__)
+        return pd.Series(indices).map(self.dummies.__getitem__)
 
 
 class LabelEncodingTransformer(BaseTransformer):

--- a/tests/unit/transformers/test_categorical.py
+++ b/tests/unit/transformers/test_categorical.py
@@ -515,6 +515,45 @@ class TestOneHotEncodingTransformer:
         expected = pd.Series(['a', 'b', 'c'])
         np.testing.assert_array_equal(out, expected)
 
+    def test_fit_dummies_no_nans(self):
+        """Test the ``fit`` method without nans.
+
+        Check that ``self.dummies`` does not
+        contain nans.
+
+        Input:
+        - Series with values
+        """
+
+        # Setup
+        ohet = OneHotEncodingTransformer()
+
+        # Run
+        data = pd.Series(['a', 2, 'c'])
+        ohet.fit(data)
+
+        # Assert
+        np.testing.assert_array_equal(ohet.dummies, ['a', 2, 'c'])
+
+    def test_fit_dummies_nans(self):
+        """Test the ``fit`` method without nans.
+
+        Check that ``self.dummies`` contain ``np.nan``.
+
+        Input:
+        - Series with values
+        """
+
+        # Setup
+        ohet = OneHotEncodingTransformer()
+
+        # Run
+        data = pd.Series(['a', 2, 'c', None])
+        ohet.fit(data)
+
+        # Assert
+        np.testing.assert_array_equal(ohet.dummies, ['a', 2, 'c', np.nan])
+
     def test_fit_no_nans(self):
         """Test the ``fit`` method without nans.
 
@@ -535,9 +574,9 @@ class TestOneHotEncodingTransformer:
 
         # Assert
         np.testing.assert_array_equal(ohet.dummies, ['a', 'b', 'c'])
-        np.testing.assert_array_equal(ohet.decoder, ['a', 'b', 'c'])
-        assert ohet.dummy_encoded
-        assert not ohet.dummy_na
+        np.testing.assert_array_equal(ohet._uniques, ['a', 'b', 'c'])
+        assert ohet._dummy_encoded
+        assert not ohet._dummy_na
 
     def test_fit_no_nans_numeric(self):
         """Test the ``fit`` method without nans.
@@ -559,9 +598,9 @@ class TestOneHotEncodingTransformer:
 
         # Assert
         np.testing.assert_array_equal(ohet.dummies, [1, 2, 3])
-        np.testing.assert_array_equal(ohet.decoder, [1, 2, 3])
-        assert not ohet.dummy_encoded
-        assert not ohet.dummy_na
+        np.testing.assert_array_equal(ohet._uniques, [1, 2, 3])
+        assert not ohet._dummy_encoded
+        assert not ohet._dummy_na
 
     def test_fit_nans(self):
         """Test the ``fit`` method with nans.
@@ -582,10 +621,10 @@ class TestOneHotEncodingTransformer:
         ohet.fit(data)
 
         # Assert
-        np.testing.assert_array_equal(ohet.dummies, ['a', 'b'])
-        np.testing.assert_array_equal(ohet.decoder, ['a', 'b', np.nan])
-        assert ohet.dummy_encoded
-        assert ohet.dummy_na
+        np.testing.assert_array_equal(ohet.dummies, ['a', 'b', np.nan])
+        np.testing.assert_array_equal(ohet._uniques, ['a', 'b'])
+        assert ohet._dummy_encoded
+        assert ohet._dummy_na
 
     def test_fit_nans_numeric(self):
         """Test the ``fit`` method with nans.
@@ -606,10 +645,10 @@ class TestOneHotEncodingTransformer:
         ohet.fit(data)
 
         # Assert
-        np.testing.assert_array_equal(ohet.dummies, [1, 2])
-        np.testing.assert_array_equal(ohet.decoder, [1, 2, np.nan])
-        assert not ohet.dummy_encoded
-        assert ohet.dummy_na
+        np.testing.assert_array_equal(ohet.dummies, [1, 2, np.nan])
+        np.testing.assert_array_equal(ohet._uniques, [1, 2])
+        assert not ohet._dummy_encoded
+        assert ohet._dummy_na
 
     def test_fit_single(self):
         # Setup
@@ -636,8 +675,8 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series(['a', 'b', 'c'])
-        ohet.dummies = ['a', 'b', 'c']
-        ohet.num_dummies = 3
+        ohet._uniques = ['a', 'b', 'c']
+        ohet._num_dummies = 3
 
         # Run
         out = ohet._transform(data)
@@ -665,10 +704,10 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series(['a', 'b', 'c'])
-        ohet.dummies = ['a', 'b', 'c']
-        ohet.indexer = [0, 1, 2]
-        ohet.num_dummies = 3
-        ohet.dummy_encoded = True
+        ohet._uniques = ['a', 'b', 'c']
+        ohet._indexer = [0, 1, 2]
+        ohet._num_dummies = 3
+        ohet._dummy_encoded = True
 
         # Run
         out = ohet._transform(data)
@@ -696,9 +735,9 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series([np.nan, None, 'a', 'b'])
-        ohet.dummies = ['a', 'b']
-        ohet.dummy_na = True
-        ohet.num_dummies = 2
+        ohet._uniques = ['a', 'b']
+        ohet._dummy_na = True
+        ohet._num_dummies = 2
 
         # Run
         out = ohet._transform(data)
@@ -728,11 +767,11 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series([np.nan, None, 'a', 'b'])
-        ohet.dummies = ['a', 'b']
-        ohet.indexer = [0, 1]
-        ohet.dummy_na = True
-        ohet.num_dummies = 2
-        ohet.dummy_encoded = True
+        ohet._uniques = ['a', 'b']
+        ohet._indexer = [0, 1]
+        ohet._dummy_na = True
+        ohet._num_dummies = 2
+        ohet._dummy_encoded = True
 
         # Run
         out = ohet._transform(data)
@@ -761,8 +800,8 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series(['a', 'a', 'a'])
-        ohet.dummies = ['a']
-        ohet.num_dummies = 1
+        ohet._uniques = ['a']
+        ohet._num_dummies = 1
 
         # Run
         out = ohet._transform(data)
@@ -791,10 +830,10 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         data = pd.Series(['a', 'a', 'a'])
-        ohet.dummies = ['a']
-        ohet.indexer = [0]
-        ohet.num_dummies = 1
-        ohet.dummy_encoded = True
+        ohet._uniques = ['a']
+        ohet._indexer = [0]
+        ohet._num_dummies = 1
+        ohet._dummy_encoded = True
 
         # Run
         out = ohet._transform(data)
@@ -822,8 +861,8 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         pd.Series(['a'])
-        ohet.dummies = ['a']
-        ohet.num_dummies = 1
+        ohet._uniques = ['a']
+        ohet._num_dummies = 1
 
         # Run
         out = ohet._transform(pd.Series(['b', 'b', 'b']))
@@ -852,9 +891,9 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         pd.Series(['a'])
-        ohet.dummies = ['a']
-        ohet.indexer = [0]
-        ohet.num_dummies = 1
+        ohet._uniques = ['a']
+        ohet._indexer = [0]
+        ohet._num_dummies = 1
         ohet.dummy_encoded = True
 
         # Run
@@ -883,9 +922,9 @@ class TestOneHotEncodingTransformer:
         # Setup
         ohet = OneHotEncodingTransformer()
         pd.Series(['a'])
-        ohet.dummies = ['a']
-        ohet.dummy_na = True
-        ohet.num_dummies = 1
+        ohet._uniques = ['a']
+        ohet._dummy_na = True
+        ohet._num_dummies = 1
 
         # Run
         out = ohet._transform(pd.Series(['b', 'b', np.nan]))
@@ -1023,7 +1062,7 @@ class TestOneHotEncodingTransformer:
         out = ohet.transform(data)
 
         # Assert
-        assert not ohet.dummy_encoded
+        assert not ohet._dummy_encoded
         np.testing.assert_array_equal(out, expected)
 
     def test_reverse_transform_no_nans(self):

